p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,153 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
+|             | can be a constructor parameter for this type ||||
+|             |---------|--------|---------|--------|
+| This type   | package | parser | control | extern |
++-------------+:-------:+:------:+:-------:+:------:+
+| package     | yes     |  no    |  no     |  no    |
+| parser      | yes     | yes    |  no     |  no    |
+| control     | yes     |  no    | yes     |  no    |
+| extern      | yes     | yes    | yes     | yes    |
+| table       |  no     |  no    |  no     |  no    |
+| value types | yes     | yes    | yes     | yes    |
+|-------------|---------|--------|---------|--------|
+
+The next table lists restrictions on where one may perform
+instantiations (see Section [#sec-instantiations]) of different types.
+The answer for `package` is always "no" because there is no
+"inside a package" where instantiations can be written in P4~16~.  One
+can definitely make constructor calls and use instances of stateful
+types as parameters when instantiating a package, and restrictions on
+those types are in the table above.
+
+For externs, one can only specify their interface in P4~16~, not their
+implementation.  Thus there is no place to instantiate objects within
+an extern.
+
+You may declare variables and constants of any of the value types
+within a parser or control.  Declaring a variable or constant is not
+the same as instantiation, hence the answer "no" in those table
+entries.  Variables may not be declared at the top level of your
+program, but constants may.
+
+|-------------|-------------------------------------------------|
+|             | can be instantiated in this place |||||
+|             |-----------|---------|---------|---------|--------|
+| This type   | top level | package | parser  | control | extern |
++-------------+:---------:+:-------:+:-------:+:-------:+:------:+
+| package     | yes       |  no     | no      | no      |  no    |
+| parser      |  no       |  no     | yes     | no      |  no    |
+| control     |  no       |  no     | no      | yes     |  no    |
+| extern      | yes       |  no     | yes     | yes     |  no    |
+| table       |  no       |  no     | no      | yes     |  no    |
+| value types |  no       |  no     | no      | no      |  no    |
+|-------------|-----------|---------|---------|---------|--------|
+
+The next table lists restrictions on what types can be passed as
+run-time parameters to other callable things that have run-time
+parameters: parsers, controls, extern methods, and actions.
+
+|-------------|-------------------------------------------------|
+|             | can be a run-time parameter to this callable thing ||||
+|             |-----------|---------|---------|---------|
+| This type   | parser    | control | method  | action  |
++-------------+:---------:+:-------:+:-------:+:-------:+
+| package     |  no       |  no     | no      | no      |
+| parser      |  no       |  no     | no      | no      |
```

Comment:
Agreed that we should support calling subparsers from parsers.  The contents of this table do not prevent that, and a different table explicitly says that parsers should be callable from parsers.

Not sure if you were aware, but p4c currently gives an error if you attempt to pass a control to another control as a run-time parameter, or a parser to another parser as a run-time parameter.  Short P4_16 programs you can try compiling to verify this can be found here: https://github.com/jafingerhut/p4-guide/tree/master/run-time-params

Maybe that is a bug in the current compiler, but keep reading below for evidence that maybe it is by design.

It does permit passing controls to other controls as _constructor_ parameters, and parser to other parsers as constructor parameters.  Is that perhaps sufficient to achieve the modularity you are hoping for?

See this relevant issue: https://github.com/p4lang/p4-spec/issues/364 which I opened partly in response to the following comment from @jnfoster, who believes that passing controls or parsers as run-time parameters should be illegal, because it enables one to write mutually recursive calls between them: https://github.com/p4lang/p4-spec/issues/361#issuecomment-318768783